pkg/vxlan/vxlan.go

```diff
@@ -134,15 +133,47 @@ func isVxlanConfigTheSame(newLink, currentLink netlink.Link) bool {
 	return true
 }
 
-func GetVtepIPAddressFrom(ipAddr string, networkPrefix int) (net.IP, error) {
-	ipSlice := strings.Split(ipAddr, ".")
-	if len(ipSlice) < 4 {
-		return nil, errors.Errorf("invalid ipAddr %q", ipAddr)
+func GetVtepIPAddressFrom(ipAddr, prefixCIDR string, family k8snet.IPFamily) (net.IP, error) {
+	ip := net.ParseIP(ipAddr)
+	if ip == nil {
+		return nil, errors.Errorf("invalid IP address: %q", ipAddr)
 	}
 
-	ipSlice[0] = strconv.Itoa(networkPrefix)
+	_, prefixNet, err := net.ParseCIDR(prefixCIDR)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse prefix %q", prefixCIDR)
+	}
+
+	switch family {
+	case k8snet.IPv4:
+		src := ip.To4()
+		prefix := prefixNet.IP.To4()
+
+		if src == nil || prefix == nil {
+			return nil, errors.Errorf("expected IPv4 addresses but got ip=%q and prefix=%q", ipAddr, prefixCIDR)
+		}
+
+		return net.IPv4(prefix[0], src[1], src[2], src[3]), nil
+	case k8snet.IPv6:
+		src := ip.To16()
+		prefix := prefixNet.IP.To16()
 
-	return net.ParseIP(strings.Join(ipSlice, ".")), nil
+		if src == nil || prefix == nil || ip.To4() != nil || prefixNet.IP.To4() != nil {
```

Comment:
Minor thing but the `To4` checks here really aren't necessary since they can't be reached unless `ip` and `prefixNet.IP` are both known to be v6.

```suggestion
		if src == nil || prefix == nil {
```